atc/exec/set_pipeline_step.go

```diff
@@ -295,6 +299,14 @@ func (s setPipelineSource) FetchPipelineConfig() (atc.Config, error) {
 		staticVars = append(staticVars, sv)
 	}
 
+	if len(s.step.plan.InstanceVars) > 0 {
+		iv := vars.StaticVariables{}
+		for k, v := range s.step.plan.InstanceVars {
+			iv[k] = v
+		}
+		staticVars = append(staticVars, iv)
+	}
```

Comment:
Precedence is kind of funny to reason about here. Does this mean that if you you set an instance var, but then set the same var through `vars`/`var_files`, the template will be evaluated with values from `vars`/`var_files` however the pipeline will be identified by the `instance_vars`?

Seems kind of sneaky. I wonder if it should be a validation warning/error. Can we think of any situation where this might occur intentionally or accidentally?

Reply:
I suppose this *could* happen intentionally if you wanted to identify the pipeline using some memorable human-understandable value while the actual underlying value is more complex. For instance, a `ref` instance var could be some friendly name, while the `ref` var could be a commit/tag SHA. Or, an `image` instance var could be an image tag, while the `image` var could be the digest.

Those aren't such compelling examples, though, since you could just as easily use a different variable name to achieve the same thing.

> Seems kind of sneaky. I wonder if it should be a validation warning/error.

Would this just be config validation, or would we want to perform actual runtime validation?

Reply:
I think it would have to be at runtime because of `var_files`. This kind of check is unprecedented (heh) though and the implementation doesn't seem like it'd be very graceful. I guess you'd have to loop over `vars` and all `var_files` to look for conflicts with `instance_vars`?

I don't feel strongly about adding the validation unless we think it'll be a "gotcha." It doesn't seem particularly high-risk, and the current behavior seems like a reasonable interpretation.